<a href="https://colab.research.google.com/github/niklasfaller/DIS21-BIg-Data/blob/main/Ex7_MNIST_classification_with_a_CNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://www.th-koeln.de/img/logo.svg" style="float: right;" width="200">

# 7th exercise: <font color="#C70039">MNIST Classification with a Convolutional Neural Network</font>
* Course: DIS21a.1
* Lecturer: <a href="https://www.gernotheisenberg.de/">Gernot Heisenberg</a>
* Author of notebook modifications and adaptations: <a href="https://www.gernotheisenberg.de/">Gernot Heisenberg</a>
* Date:   08.08.2023

<img src="https://miro.medium.com/max/1250/1*vkQ0hXDaQv57sALXAJquxA.jpeg" style="float: center;" width="600">

---------------------------------
**GENERAL NOTE 1**:
Please make sure you are reading the entire notebook, since it contains a lot of information about your tasks (e.g. regarding the set of certain paramaters or specific computational tricks, etc.), and the written mark downs as well as comments contain a lot of information on how things work together as a whole.

**GENERAL NOTE 2**:
* Please, when commenting source code, just use English language only.
* When describing an observation (for instance, after you have run through your test plan) you may use German language.
This applies to all exercises in DIS 21a.1.  

---------------------

### <font color="ce33ff">DESCRIPTION</font>:

This notebook allows you for learning how you effectively setup and use convolutional neural networks (CNN).
For this purpose, the classification of the MNIST digits from the earlier exercise is done again.
Using a densely (fully) connected ANN in that previous exercise, you have achieved a test accuracy of about 97%.
Even though the CNN will be very basic, its accuracy will blow out that of a densely-connected ANN.

This notebook shows in a few lines of code what a basic CNN looks alike.
Basically, it is a simple stack of `Conv2D` and `MaxPooling2D` layers.

-------------------------------------------------------------------------------------------------------------

### <font color="FFC300">TASKS</font>:
Within this notebook, the tasks that you need to work on are always listed as bullet points below.
If a task is more challenging and consists of several steps, this is indicated as well.
Make sure you have worked down the task list and commented your doings.
This should be done using markdown.<br>
<font color=red>Make sure you don't forget to specify your name and your matriculation number in the notebook before submitting it.</font>

**YOUR TASKS in this exercise are as follows**:
1. import the notebook to Google Colab.
2. make sure you specified you name and your matriculation number in the header below my name and date.
    * set the date too and remove mine.
3. read the entire notebook carefully.
    * for better understanding, add comments whereever you feel it necessary.
    * run the notebook for the first time and note the result in a markdown table.
        * I have provided you with an example of a table in markdown (see below). Make sure you adapt your table accordingly.
        * Put the table at the end of the notebook.
        * This type of table will be needed in the other exercises as well. Always put it at the end.
    
| type of method | loss function | optimizer | accuracy |
| :-: | :-: | :-: | :-: |
| classification | categorical_crossentropy | bamm !|.666

4. write the code for 'loading and preparing the MNIST data set'.
5. write the code for 'training the CNN'.
    * epochs=5
    * batch_size=64
6. write the code for 'evaluating the CNN', initially with the given hyperparameters.

7. Compare the obtained result to the one you obtained by the densely connected ANN and compute the increase of accuracy in percent.

8. write a test plan for testing other hyperparameters. Store the values (also the one from task 7) in a table as the one from above.

-----------------------------------------------------------------------------------

## START OF THE NOTEBOOK CODE
----------------------------------------------------------------------------------------------------------------------
### necessary imports
others are going to be included as soon as they are needed

In [1]:
import tensorflow
#tensorflow.keras.__version__

from tensorflow.keras import layers
from tensorflow.keras import models

### building the CNN

<font color="#C70039">NOTE: </font>
A CNN takes as input tensors of shape `(image_height, image_width, image_channels)`
(not including the batch dimension).

In this case here, the CNN is configured to process inputs of size `(28, 28, 1)`, which is the format of MNIST grayscale images. This is done by passing the argument `input_shape=(28, 28, 1)` to the first layer.

In [2]:
model = models.Sequential()

model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))

# the subsequent layers are taking the shape of the previous layers as input shape.
# Therefore an explicit specification of the input shape is not needed.
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))

Display the architecture of the CNN

In [3]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2  (None, 13, 13, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 conv2d_2 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
Total params: 55744 (217.75 KB)
Trainable params: 55744 

The output of every `Conv2D` and `MaxPooling2D` layer is a 3D tensor of shape `(height, width, channels)`. The width
and height dimensions tend to shrink by going deeper into the network.
The number of channels is controlled by the first argument passed to the `Conv2D` layers (e.g. 32 or 64).

The next step is to feed the last output tensor (of shape `(3, 3, 64)`) into a densely (fully) connected classifier network, i.e. a stack of `Dense` ANN layers.
These classifiers do process vectors, which are 1D, whereas our current output is a 3D tensor.

So, first the 3D outputs needs to be flattened to 1D, and then a few `Dense` layers are added on top.

In [4]:
model.add(layers.Flatten()) # this could also be done by the reshaping function. You can try it out, if you like.
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

Since this is a classification of 10 digits again, use a final layer with 10 outputs and a softmax activation (as done before already).
Now here's what the final CNN looks like.

In [5]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2  (None, 13, 13, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 conv2d_2 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 576)               0

Note, that the `(3, 3, 64)` outputs were flattened into vectors of shape `(576,)`, before going through two `Dense` layers.

Now, train the CNN on the MNIST digits.

### loading and preparing the MNIST data set
as done before (compare with your old code)

In [6]:
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

In [7]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


11490434/11490434 [==============================] - 1s 0us/step


### training the CNN

for the start, take
* epochs=5
* batch_size=64

In [8]:
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images,train_labels,epochs=5,batch_size=64)


Epoch 1/5
938/938 [==============================] - 53s 55ms/step - loss: 0.1711 - accuracy: 0.9471
Epoch 2/5
938/938 [==============================] - 50s 53ms/step - loss: 0.0469 - accuracy: 0.9852
Epoch 3/5
938/938 [==============================] - 51s 54ms/step - loss: 0.0318 - accuracy: 0.9901
Epoch 4/5
938/938 [==============================] - 49s 53ms/step - loss: 0.0244 - accuracy: 0.9926
Epoch 5/5
938/938 [==============================] - 51s 54ms/step - loss: 0.0183 - accuracy: 0.9940


### evaluate the CNN
Evaluate the model on the test data.

In [9]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 [==============================] - 3s 9ms/step - loss: 0.0314 - accuracy: 0.9915


In [10]:
test_acc

0.9915000200271606

While the densely connected ANN had a test accuracy of about 97,85%.

The basic CNN has a test accuracy of about 99,15% (double bamm !!!).

This is an increase of 1,3 %

### <font color="#C70039">Include your result table here and reflect a good test plan (see task list)</font>

In [13]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images,train_labels,epochs=5,batch_size=32)

test_loss, test_acc = model.evaluate(test_images, test_labels)

Epoch 1/5
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0099 - accuracy: 0.9972
Epoch 2/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0092 - accuracy: 0.9974
Epoch 3/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0072 - accuracy: 0.9977
Epoch 4/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0070 - accuracy: 0.9979
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 0.0500 - accuracy: 0.9935


In [14]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images,train_labels,epochs=10,batch_size=64)

test_loss, test_acc = model.evaluate(test_images, test_labels)

Epoch 1/10
938/938 [==============================] - 52s 54ms/step - loss: 0.0042 - accuracy: 0.9987
Epoch 2/10
938/938 [==============================] - 50s 54ms/step - loss: 0.0027 - accuracy: 0.9991
Epoch 3/10
938/938 [==============================] - 49s 53ms/step - loss: 0.0026 - accuracy: 0.9993
Epoch 4/10
938/938 [==============================] - 51s 54ms/step - loss: 0.0027 - accuracy: 0.9993
Epoch 5/10
938/938 [==============================] - 49s 52ms/step - loss: 0.0018 - accuracy: 0.9995
Epoch 6/10
938/938 [==============================] - 51s 54ms/step - loss: 0.0023 - accuracy: 0.9994
Epoch 7/10
938/938 [==============================] - 49s 52ms/step - loss: 0.0011 - accuracy: 0.9997
Epoch 8/10
938/938 [==============================] - 51s 55ms/step - loss: 0.0016 - accuracy: 0.9995
Epoch 9/10
938/938 [==============================] - 50s 53ms/step - loss: 0.0017 - accuracy: 0.9995
Epoch 10/10
313/313 [==============================] - 3s 9ms/step - loss: 0.0679 

| type of network | type of method | loss_function | optimizer | batch_size | epochs | accuracy |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| ANN | classification | categorical_crossentropy | rmsprop | 128 | 5 | 0.9785
| CNN | classification | categorical_crossentropy | rmsprop | 32 | 5 | 0.9935
| CNN | classification | categorical_crossentropy | rmsprop | 64 | 10 | 0.9926